### PR TITLE
MGMT-12423: Watch for Agent changes in ClusterDeployment

### PR DESCRIPTION
**Pull request by mkowalski**
In order to catch cases where changes to the Agent CR affect AgentClusterInstall but they do not go over the AI DB, we are adding a watch over Agents that reference the ACI in subject.

Because it's ClusterDeployment reconciler that in reality modifies the ACI CR, we are placing a watch in the CD CR and not the ACI CR.

Contributes-to: [MGMT-12423](https://issues.redhat.com//browse/MGMT-12423)

/cc @tsorya

### CODE REVIEW

**Comment by filanov** [comment]
/hold 
can we check the performance for this change? 
And as Mat suggested maybe we could look into other solutions? https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/

**Comment by filanov** [comment]
/assign @mhrivnak 
@mhrivnak what do you think watch vs event filtering?

**Comment by filanov** [comment]
Once it's merged lets backport into 2.6 we saw a similar case in a subsystem that failed when agent cluster install didn't catch this sync change in the agent 
cc: @slaviered

**Comment by mhrivnak** [comment]
Can you provide an example of an Agent status change that requires reconciling the ClusterDeployment?

**Review by nmagnezi — CHANGES_REQUESTED** [review]
What about adding tests to ensure it behaves as expected?
I think we need both UTs and subsystem tests.

**Comment by mkowalski** [comment]
> Can you provide an example of an Agent status change that requires reconciling the ClusterDeployment?

Sure thing. Imagine the following steps

1) you create Agent with no ignition overrides; Agent has SpecSync OK and ClusterDeployment is Ready
2) you modify Agent by adding ignition override which has a wrong syntax; Agent goes into SpecSync Error

After step (2) you would expect ClusterDeployment to go into NotReady with the reason "You have 1 Agent with Error". This is not happening because change in Agent never triggered reconciliation. With this PR I am submitting here, a change in Agent's SpecSync will trigger a reconciliation.

**Comment by mkowalski** [comment]
> What about adding tests to ensure it behaves as expected? I think we need both UTs and subsystem tests.

I agree 100% with this, but given this is some amount of work, I need to first have an agreement that the direction I take here (i.e. Watch with Predicate) is an approved design. Once this is settled, I will make this PR a proper piece of code with tests&co.

**Comment by slaviered** [comment]
We already have a subsystem test that can benefit from this logic because it has a race condition with agent changes and aci changes: see '`deploy clusterDeployment with agent and invalid ignition config`'

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (paul-maidment) [inline internal/controller/controllers/clusterdeployments_controller.go:1556]: I'm guessing the reason we don't warn about the lack of `agent.Spec.ClusterDeploymentName` in the log is that this is common and to report on it would generate too much noise?

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (paul-maidment) [inline internal/controller/controllers/clusterdeployments_controller.go:1564]: Should we log anything if we are having trouble obtaining the agent objects?

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (paul-maidment) [inline internal/controller/controllers/clusterdeployments_controller.go:1569]: Should we log anything if we are having trouble obtaining the agent objects?

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (mkowalski) [reply internal/controller/controllers/clusterdeployments_controller.go:1556]: We don't warn about it, because this is not really a reason for a warning. Agent without a respective ClusterDeployment is still a valid agent (think about late-binding or some other scenarios).

We could at most `log.Debug` about it, but I don't personally see a big benefit

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (mkowalski) [reply internal/controller/controllers/clusterdeployments_controller.go:1564]: Nope, because this is not a trouble *obtaining* the object, it would be a trouble with converting the type.

What we do here is we receive `updateEvent.ObjectOld` which we need to map into Agent object. It's purely implementation detail that we need this cast, because even though `updateEvent.ObjectOld` comes from the Watch over `aiv1beta1.Agent`, it doesn't have its type. That's because `updateEvent` itself has `event.UpdateEvent` type. We would need generics or something to have it more (or less) elegant, but as this comes from the framework, I just consume it "as is"

File: internal/controller/controllers/clusterdeployments_controller.go
Comment (mkowalski) [reply internal/controller/controllers/clusterdeployments_controller.go:1569]: As above, no because it's a framework

**Review by filanov — APPROVED** [review]

**Comment by mkowalski** [comment]
/unhold

Trivial to revert if any perf issues are observed

**Pull request merged**
